lib/ed25519.js

```diff
@@ -36,6 +44,165 @@ ed25519.constants.SEED_BYTE_LENGTH = 32;
+  type: asn1.Type.SEQUENCE,
+  constructed: true,
+  composed: true,
+  value: [{
+    tagClass: asn1.Class.UNIVERSAL,
+    type: asn1.Type.SEQUENCE,
+    constructed: true,
+    composed: true,
+    value: [{
+      tagClass: asn1.Class.UNIVERSAL,
+      type: asn1.Type.OID,
+      constructed: false,
+      composed: false,
+      capture: 'ed25519Oid'
+    }]
+  }, {
+    tagClass: asn1.Class.UNIVERSAL,
+    type: asn1.Type.BITSTRING,
+    constructed: false,
+    composed: true,
+    captureBitStringValue: 'ed25519PublicKey'
+  }]
+};
+
+/**
+ * ASN1 private key validator for importing ed25519 keys in DER format.
+ * @type {Object}
+*/
+var ed25519PrivateKeyValidator = forge.ed25519.privateKeyValidator = {
+  name: 'ed25519PrivateKey',
+  tagClass: asn1.Class.UNIVERSAL,
+  type: asn1.Type.SEQUENCE,
+  constructed: true,
+  composed: true,
+  value: [{
+    tagClass: asn1.Class.UNIVERSAL,
+    type: asn1.Type.INTEGER,
+    constructed: false,
+    composed: false,
+    capture: 'privateKeyZero' // This value should always be 0.
+  }, {
+    tagClass: asn1.Class.UNIVERSAL,
+    type: asn1.Type.SEQUENCE,
+    constructed: true,
+    composed: true,
+    value: [{
+      tagClass: asn1.Class.UNIVERSAL,
+      type: asn1.Type.OID,
+      constructed: false,
+      composed: false,
+      capture: 'ed25519Oid'
+    }]
+  }, {
+    tagClass: asn1.Class.UNIVERSAL,
+    type: asn1.Type.OCTETSTRING,
+    constructed: false,
+    composed: false,
+    capture: 'ed25519PrivateKey'
+  }]
+};
+
+/**
+ * Used to extract the asn1 from DER formatted key.
+ *
+ * @param {NativeBuffer} keyBuffer - A DER formatted key.
+ *
+ * @return {Object} An ASN1 object.
+ */
+function getKeyBytes(keyBuffer) {
+  var binary = forge.util.binary.raw.encode(keyBuffer);
+  return forge.asn1.fromDer(binary);
+}
+
+/**
+ * Given a DER formatted key this function will extract the raw key value.
+ *
+ * @param {Object} options - The options for this operation.
+ * @param {NativeBuffer} options.keyBuffer -
+ * The result of exporting the DER key.
+ * @param {ed25519PrivateKeyValidator|ed25519PublicKeyValidator}
+ * options.validator - A validator to check the exported key.
+ * @param {Object} [options.capture = {}] - A collection of name value pairs
+ * captured by the validator.
+ * @param {'ed25519PrivateKey'|'ed25519PublicKey'} options.keyName -
+ * The name of the key captured by the validator.
+ *
+ * @returns {string} - A binary string or the key value.
+ */
+ed25519.extractKey = function(options) {
+  var capture = options.capture || {};
+  var asn1 = getKeyBytes(options.keyBuffer);
+  var errors = [];
+  var valid = forge.asn1.validate(asn1, options.validator, capture, errors);
+  if(!valid) {
+    throw new Error('Invalid Key: ' + errors.join('.'));
+  }
+  var oid = forge.asn1.derToOid(capture.ed25519Oid);
+  var ed25519Oid = forge.oids.EdDSA25519;
+  if(oid !== ed25519Oid) {
+    throw new Error('Invalid OID expected ' + ed25519Oid +
+      ' recieved ' + oid);
```

Comment:
Typo here on "recieved" but let's rewrite this error message to be more in line with others:

`'Invalid OID "' + oid + '"; OID must be "' + ed25519Oid + '".'`;